src/Groups/matrices/iso_oscar_gap.jl

```diff
@@ -1,78 +1,110 @@
 
-function gen_ring_iso(F::FqNmodFiniteField)
-   p = Int64(characteristic(F))
-   d = Int64(degree(F))
-   z = gen(F)
+      # For "small" primes x should be an FFE, but for bigger ones it's GAP_jll.Mptr (?)
+      finv1(x) = F(GAP.gap_to_julia(GAP.Globals.IntFFE(x)))
 
-   if d==1
-      myOne = GAP.Globals.One(GAP.Globals.GF(p))
-      f(x::fq_nmod) = Int(coeff(x,0))*myOne
-      finv(x::FFE) = F(GAP.Globals.IntFFE(x))
-      return GenRingIso(F, GAP.Globals.GF(p), f, finv)
+      return GenRingIso{T, elem_type(F)}(F, Fp_gap, f1, finv1)
    end
-   L = [Int64(lift(coeff(defining_polynomial(F),i))) for i in 0:d-1]
-   L = vcat(L,[1])
-   i = GAP.Globals.One(GAP.Globals.GF(p))
-   L_gap = GAP.julia_to_gap([i*y for y in L])
-   f_gap = GAP.Globals.UnivariatePolynomial(GAP.Globals.GF(p),L_gap)
-   F_gap = GAP.Globals.GF(GAP.Globals.GF(p),f_gap)
-   Basis_F = GAP.Globals.Basis(F_gap)
-   Basis_g_j = GAP.gap_to_julia(GAP.Globals.BasisVectors(Basis_F))
-   homom(x::fq_nmod) = elem_f(x,Basis_g_j,d)
-   homominv(x::FFE) = elem_g(x, Basis_F, z, d)
-   return GenRingIso(F, F_gap, homom, homominv)
-end
 
-(g::GenRingIso)(x::fq_nmod) = g.f(x)
-(g::GenRingIso)(x::FFE) = g.f_inv(x)
+   e = GAP.Globals.One(Fp_gap)
+   L = [ GAP.julia_to_gap(lift(coeff(defining_polynomial(F), i)))*e for i in 0:d - 1 ]
+   push!(L, e)
+   L_gap = GAP.julia_to_gap(L)
+   f_gap = GAP.Globals.UnivariatePolynomial(Fp_gap, L_gap)
+   G = GAP.Globals.GF(Fp_gap, f_gap)
+
+   # TODO: Somehow Basis does not work for some (?) field. Should it or is there
+   # a better way (on the GAP side)?
+   # For example
+   # > F, a = FiniteField(next_prime(2^8), 2, "a")
+   # does not work.
+   Basis_G = GAP.Globals.Basis(G)
+   Basis_F = Vector{elem_type(F)}(undef, d)
+   Basis_F[1] = F(1)
+   for i = 2:d
+      Basis_F[i] = Basis_F[i - 1]*gen(F)
+   end
 
-Base.show(io::IO, f::GenRingIso) = print(io, "Ring isomorphism between ", f.domain, " and the corresponding GAP")
+   function f(x::Union{fq_nmod, fq})
+      v = [ GAP.julia_to_gap(coeff(x, i)) for i in 0:d - 1 ]
+      return sum([ v[i]*Basis_G[i] for i in 1:d ])
+   end
+
+   # For "small" primes x should be an FFE, but for bigger ones it's GAP_jll.Mptr (?)
+   function finv(x)
+      v = GAP.Globals.Coefficients(Basis_G, x)
+      # TODO: IntFFE apparently does not work for "big" primes, e.g.
+      # > F, a = FiniteField(next_prime(fmpz(2)^64), 2, "a")
+      # does not work.
+      v_int = [ GAP.gap_to_julia(GAP.Globals.IntFFE(v[i])) for i = 1:length(v) ]
+      return sum([ v_int[i]*Basis_F[i] for i = 1:d ])
+   end
+
+   return GenRingIso{T, elem_type(F)}(F, G, f, finv)
+end
 
+function (g::GenRingIso{T, S})(x::S) where { T, S }
+   @assert parent(x) === g.domain "Not an element of the domain"
+   return g.f(x)
+end
+
+# TODO: Do we really want it like this? I does not make any sense intuitively
+# (to me) to call the map itself for the preimage/inverse.
+(g::GenRingIso)(x) = g.f_inv(x)
```

Comment:
Well, for `fq_nmod` etc there is `(g::GenRingIso{T, S})(x::S)` in line 98. But I agree of course that it should not be like this. We can't just remove it since this is used elsewhere. I'll look into it.

Reply:
So I completely removed those ad hoc types `GenRingIso` and `GenMatIso` and use maps of generic type `MapFromFunc` instead. I hope I found all the places where `f(a)` is called, but `preimage(f, a)` is meant.